### PR TITLE
Feature osom converter

### PR DESCRIPTION
**Pull request by kmilo9999**

### CODE REVIEW

**Comment by kmilo9999** [comment]
@mirestrepo @gtdang  a quick review on this.

File: tiff_2_png/tiff_2_png_converter.py [hunk 1]
```diff
@@ -235,33 +239,56 @@ def convert_to_png(
...
         # export as single channel image
-        file_name = name + "_chn_" + str(channel) + extension
-        logging.info("##Saving channel " + str(channel) + " Image to: " + file_name)
-        cv2.imwrite(file_name, image_out)
-        file = open(name + "_metadata", "a")
+        file_name = file_name + "_chn_" + str(channel)
+        file_name_full_path = os.path.join(parent_folder, file_name + extension)
+        logging.info(
+            "##Saving channel " + str(channel) + " Image to: " + file_name_full_path
+        )
```
Comment (gtdang) [inline tiff_2_png/tiff_2_png_converter.py:265]: Cleaner to use f string
```suggestion
        logging.info(
            f"##Saving channel {channel} image to: {file_name_full_path}"
        )
```

**Review by gtdang — COMMENTED** [review]
Did a quick look-over. Had some logic that I think needs checking and some minor string formatting nit-picks.

File: osom_converter/create_volume_viewer_data.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    query_depths = np.arange(min_z, max_z + 1)
+    query_depths_shape = np.shape(query_depths)
+    slices = query_depths_shape[0]
+
+    # getting minimum and max data
+    min_data = np.floor(data.min())
+    max_data = np.ceil(data.max())
+
+    # scale data to the range of 0 and 1
+    data = (data - min_data) / (max_data - min_data)
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply osom_converter/create_volume_viewer_data.py:82]: The logic works me.  `data` is between [0,1]. Ceiling and floor operations are to get a close interval without high number decimals and/or values over 1. Let me know if that answers the question.

File: tiff_2_png/tiff_2_png_converter.py [hunk 3]
```diff
@@ -222,11 +222,15 @@ def convert_to_png(
...
                     else:
                         images_in_sequence[slice] = img
-                    max_pixel = (max_pixel, np.max(img))[np.max(img) > max_pixel]
-                    min_pixel = (min_pixel, np.min(img))[np.min(img) < min_pixel]
+                    global_max_pixel = (global_max_pixel, np.max(img))[
+                        np.max(img) > global_max_pixel
+                    ]
+                    global_min_pixel = (global_min_pixel, np.min(img))[
+                        np.min(img) < global_min_pixel
+                    ]
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply tiff_2_png/tiff_2_png_converter.py:230]: That is python ternary operator.

File: osom_converter/create_volume_viewer_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gtdang) [reply osom_converter/create_volume_viewer_data.py:82]: Just to make sure I understand correctly... The data before before it's overwritten is floating point array. Say that max of the data is 9.2. Ceil will make `max_data` = 10 and that value is used in the scaling. Therefore the upper endpoint of the scaled data is scaled to the value 10 and not the true max of the data (9.2). This would also happen for the minimum end of the range. Just wanted to make sure that difference was trivial. Does that make sense?

File: osom_converter/create_volume_viewer_data.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    stream.write(format_spec % args)
+
+
+@app.command()
+def createOsomData(
+    osom_gridfile: str, osom_data_file: str, output_folder: str, data_descriptor: str
+):
+
+    """
+    Converts NETCDF files (*.nc) form the osom model to data files that can be read
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply osom_converter/create_volume_viewer_data.py:21]: Done

File: osom_converter/create_volume_viewer_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kmilo9999) [reply osom_converter/create_volume_viewer_data.py:82]: I see your point, and until now I am aware of this detail. Graphically, it would mean that  no pixel will have the maximum color intensity ( i.e color space 8 bit = 255 ). Taking the floor of the maximum would mean that the most intense pixel is not the real maximum of the dataset.  
Let's say the maximum value is 9.238456 and the prior number close to the maximum is 9.23333. Taking the floor of the maximum would make us lose those two values and all of the other numbers between the range [ 9 ,  9.238456 ]. Taking the maximum's ceiling includes the values in that range, but again the maximum pixel intensity in the frame would be 248 (in the range 0-255).
TLDR:  Taking the maximum's ceiling can give us small graphic artifacts, but at least we wont be losing data.

File: osom_converter/create_volume_viewer_data.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    OsomDataFilePath, osom_data_filename = os.path.split(osom_data_file)
+    osom_data_filename, ext = os.path.splitext(osom_data_filename)
+    if not os.path.exists(output_folder):
+        os.mkdir(output_folder)
+
+    output_f = os.path.join(output_folder, "data")
+    if not os.path.exists(output_f):
+        os.mkdir(output_f)
+
+    start_time = datetime.strptime("2006-01-01 00:00:00", "%Y-%m-%d %H:%M:%S")
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply osom_converter/create_volume_viewer_data.py:93]: I added code to extract the real ocean time from the original nc file.

File: tiff_2_png/tiff_2_png_converter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmilo9999) [reply tiff_2_png/tiff_2_png_converter.py:230]: Actually, I am gonna change this part of the code to avoid further confusions.

File: tiff_2_png/tiff_2_png_converter.py [hunk 6]
```diff
@@ -222,11 +222,15 @@ def convert_to_png(
...
 
-        logging.info("MAX Value in volume: " + str(max_pixel))
-        logging.info("MIN Value in volume: " + str(min_pixel))
+        logging.info("MAX Value in volume: " + str(global_max_pixel))
+        logging.info("MIN Value in volume: " + str(global_min_pixel))
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply tiff_2_png/tiff_2_png_converter.py:233]: Done

File: tiff_2_png/tiff_2_png_converter.py (hunk 1 above)
Comment (kmilo9999) [reply tiff_2_png/tiff_2_png_converter.py:265]: I like this kind of format. It's better than adding strings.

File: osom_converter/create_volume_viewer_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gtdang) [reply osom_converter/create_volume_viewer_data.py:94]: Great! Yeah that makes sense.

**Comment by kmilo9999** [comment]
@gtdang quick review on this PR

**Review by gtdang — COMMENTED** [review]
Did a quick review mostly of the tiff 2 png converter. Have some comments about global mins/max.

File: tiff_2_png/tiff_2_png_converter.py [hunk 7]
```diff
@@ -235,33 +234,54 @@ def convert_to_png(
...
+        file_name, extension = os.path.splitext(file_base_name)
         if not extension:
             extension = ".png"
 
         # export as single channel image
-        file_name = name + "_chn_" + str(channel) + extension
-        logging.info("##Saving channel " + str(channel) + " Image to: " + file_name)
-        cv2.imwrite(file_name, image_out)
-        file = open(name + "_metadata", "a")
+        file_name = file_name + "_chn_" + str(channel)
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply tiff_2_png/tiff_2_png_converter.py:256]: done!

File: tiff_2_png/tiff_2_png_converter.py [hunk 8]
```diff
@@ -207,8 +206,8 @@ def convert_to_png(
                     ## check all images have the same data type
                     if bits_per_sample == "":
                         bits_per_sample = img.dtype
-                        max_pixel = np.iinfo(img.dtype).min
-                        min_pixel = np.iinfo(img.dtype).max
+                        global_max_pixel = np.iinfo(img.dtype).min
+                        global_min_pixel = np.iinfo(img.dtype).max
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply tiff_2_png/tiff_2_png_converter.py:210]: I changed the logic in this section of the code. The first iteration takes the min and max of the first image and in the next one compares the current image against these values.

File: tiff_2_png/tiff_2_png_converter.py [hunk 9]
```diff
@@ -222,11 +221,11 @@ def convert_to_png(
                         images_in_sequence[slice] = img[channel]
                     else:
                         images_in_sequence[slice] = img
-                    max_pixel = (max_pixel, np.max(img))[np.max(img) > max_pixel]
-                    min_pixel = (min_pixel, np.min(img))[np.min(img) < min_pixel]
+                    global_max_pixel = max(global_max_pixel, np.max(img))
+                    global_min_pixel = min(global_min_pixel, np.min(img))
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply tiff_2_png/tiff_2_png_converter.py:225]: Check out the new code. Maybe it's more clear now.

[the remark this answers is not in the document]
Comment (gtdang) [reply tiff_2_png/tiff_2_png_converter.py:225]: Ah ok, that logic makes sense. Is there a reason to initialize the global min/max with the data type machine limits on lines 192-193?

**Review by gtdang — COMMENTED** [review]
Looks good. Just one question.

File: tiff_2_png/tiff_2_png_converter.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kmilo9999) [reply tiff_2_png/tiff_2_png_converter.py:225]: Not specific reason. Just finding a big and small number to initialize those variables.
